cronjobs/unknown-vms.sh

```diff
@@ -5,64 +5,62 @@ set -exu
 
-#function to build JMESpath query to exclude certain hosts based on ip
+# build JMESpath filter to exclude a list of instances based on their PrivateIpAddress
 query_filter() {
     local IFS
     unset IFS
-    local QUERY=""
+    local FILTER=""
 
     for ip in ${1}; do
-        if [ -z "$QUERY" ]; then
-            QUERY="?PrivateIpAddress != "
+        if [ -z "$FILTER" ]; then
+            FILTER="?PrivateIpAddress != "
         else
-            QUERY="${QUERY} && PrivateIpAddress != "
+            FILTER="${FILTER} && PrivateIpAddress != "
         fi
-        QUERY="${QUERY}\`$ip\`"
+        FILTER="${FILTER}\`$ip\`"
     done
 
-    QUERY="Reservations[].Instances[${QUERY}] | "
-
-    echo ${QUERY}
+    echo ${FILTER}
 }
 
 # get a list of all the instances bosh has created
 KNOWN_INSTANCES=$($PSQL -h ${PGHOST} -U ${PGUSERNAME} -d ${PGDBNAME} -t -c "select uuid from instances")
 
-# find the VPC ID we want to enumerate
-VPCID=$(${AWSCLI} ec2 describe-vpcs --filter Name=tag:Name,Values=${VPC_NAME} --output text --query 'Vpcs[].VpcId')
-
-# find all instances in that VPC and emit a metric for each unknown host
-for id in $(${AWSCLI} ec2 describe-instances --max-items 500 --output text  --filter Name=vpc-id,Values=${VPCID} --query "$(query_filter "${BOSH_DIRECTOR} ${INSTANCE_WHITELIST}") [].{\"aws_id\": InstanceId, \"bosh_id\": Tags[?Key==\`id\`].Value | [0]} | [].[bosh_id || aws_id]"); do
-    STATE="ok"
-
-    if [[ $KNOWN_INSTANCES != *${id}* ]]; then
-        STATE="critical"
+# find the AWS VPC ID we want to enumerate
+VPC_ID=$(${AWSCLI} ec2 describe-vpcs --filter Name=tag:Name,Values=${VPC_NAME} --output text --query 'Vpcs[].VpcId')
+
+# emit a metric for all instances in that VPC
+IFS="\n"
+for vminfo in $(
+        ${AWSCLI} ec2 describe-instances --max-items 1000 --output text  --filter Name=vpc-id,Values=${VPC_ID} --query "
+            Reservations[].Instances[$(query_filter "${BOSH_DIRECTOR} ${INSTANCE_WHITELIST}")]
+            | [].{\"launch\": LaunchTime, \"iaas_id\": InstanceId, \"bosh_id\": Tags[?Key==\`id\`].Value | [0]}
+            | [].[iaas_id, bosh_id, launch]"
+        )
+    do
+
+    iaas_id=$(echo ${vminfo} | cut -f1)
+    bosh_id=$(echo ${vminfo} | cut -f2)
+    uptime=$(expr $(date +'%s') - $(date -d "$(echo ${vminfo} | cut -f3)" + '%s'))
+
+    # check to see if bosh director knows about this instance pulled from the iaas
+    unknown_instance=0
+    if [[ $KNOWN_INSTANCES != *${bosh_id}* ]]; then
+        unknown_instance=1
     fi
 
-    #${RIEMANNC} --service "unknown-vm.found" --host ${id} --state ${STATE} --ttl 120 --metric_sint64 1
-
-    cat <<EOF | curl --data-binary @- "${GATEWAY_HOST}:${GATEWAY_PORT:-9091}/metrics/job/awslogs/instance/${NICE_GROUP}"
-    awslogs_loggroup_not_logging {group="${NICE_GROUP}"} ${STATUS}
+    cat <<EOF | curl --data-binary @- "${GATEWAY_HOST}:${GATEWAY_PORT:-9091}/metrics/job/bosh/unknowninstance/${VPC_NAME}/${iaas_id}"
```

Comment:
Is this a valid push gateway url? Looks like push gateway listens at `"/metrics/job/:job/*labels"`.

https://github.com/prometheus/pushgateway/blob/master/main.go#L73

Reply:
Also, could we send all of these metrics in a single `PUT`?